jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala

```diff
@@ -294,6 +297,32 @@ object XGBoost extends Serializable {
     }
   }
 
+  private def repartitionForTrainingGroup(
+      trainingData: RDD[XGBLabeledPoint], nWorkers: Int): RDD[Array[XGBLabeledPoint]] = {
+    val normalGroups = trainingData.mapPartitions(
+      new LabeledPointGroupIterator(_)).filter(!_._1).map(_._2)
+    val edgeGroups = trainingData.mapPartitions(
+      new LabeledPointGroupIterator(_)).filter(_._1).map(r => (TaskContext.getPartitionId(), r._2))
+    // the edgeGroups should be small enough to be processed in one partition
+    val stitchedGroup = edgeGroups.coalesce(1)
```

Comment:
Let's do some math here.

Assume that we have a 100GB training data, the HDFS block size is 128MB, we have 800 partitions when loading data into RDD.
for each RDD partition, there are two edge groups, the first group and the last group, so the total number of edge groups are 800 * 2, which is small enough to be coalesced to one executor and processed.

What is an edge group? See if a group has 100 LabeledPoint, 30 of them may in one RDD partition, the 70 of them may in the next RDD partition. So there are only two edge groups in each partition.

The millions of groups other then the first and last groups in a RDD partition are *normal* groups and we do not need to stitch them.

Reply:
I just use HDFS as an example, since we only collect two groups from each partition, the total size of edge groups is relatively small. So this is actually a optimization for performance, not causing performance issue.

Or we can use RDD group by which actually does the same logic but in a distributed fashion.

Reply:
you still didn’t answer my previous question, how about the user pass in a RDD containing 4 partitions each of which contains only 1/2 group data, 

All groups will be collected into a single partition?

Reply:
Yes, all data are collection into one partition, and then produces a list of groups in one partition.